### ISSUE TITLE
Cancel push on API error

### ISSUE THREAD

**Issue by canpolat** [odeke-em/drive#519]
When a folder is pushed without prompt, drive should have an option to stop transfer if there is an error during `push`.

Reason:

I was testing a script that I will use to sync my local files to cloud. The script contains the following line:

    drive push -quiet

I ran this script a couple of times to test it. Everything was sync and I got:

    Resolving...
    Everything is up-to-date.

All good. Then I made a change in the script and re-ran it. The result was scary (remember there were no new files):

    Resolving...
    googleapi: Error 503: Backend Error, backendError
    36163636 / 1284737695
    [===>-- ... --------------------------------------------------------] 2.81 % 37m11s

I stopped the script immediately. Then re-ran it to see the result (which was even scarier):

    Resolving...
    These paths clash
    ... (list of files here)
    clashes detected. Use `ignore-name-clashes` to override this behavior or `fix-clashes` to try fixing this

In short, if I was running this without the `-quiet` flag, I would have noticed that there is something fishy (when this error occurs, drive keeps the computation and as a result, reports many files not being synched to cloud) and would not allow the push.

My suggestion: When running `drive push` in `-quiet` mode the user should have an option to bail out if there is an error when resolving the diff.

Also: Probably, the resolving algorithm should be fixed so that it doesn't continue if there is an API error.

**Comment by odeke-em** [comment]
Hello @canpolat thank you for reporting this.

I see, so you want a "preceeding all pass else fail entirely mode"? Google Drive's backend occasionally trips out hence the 503's and since that is retryable it is automatically retried for you unless it constantly fails and timeouts out after >16 tries. Yeah the clashes were probably created during that period of a concurrent push when Google Drive's backend failed for one then was restarted for another goroutine e.g
`/example/b1/b2 and /example/c1/c2` when that error occurs if both routines don't lock on common dir `/example` the clashes will creep in. 
Thank you for the detail in this and I'll jump into this at the end of next week once this semester's exams are done. For now if you think something tricky is going on, please use `--verbose` and then you'll see the respective fails when they happen.

**Comment by canpolat** [comment]
@odeke-em, I experienced the 503 error only twice (on different days on different machines). And in both cases clashes were created. Good thing is, the `-fix-clashes` flag fixes them. As you may guess, it's not easy to reproduce :)

I think this option makes more sense when pushing `quiet`ly; but having it as a general option to `push` and `pull` would help preventing human error.

When it comes to the fix: No rush. Good luck with your exams.

**Comment by gauteh** [comment]
Canpolat writes on December 9, 2015 19:23:
> My suggestion: When running `drive push` in `-quiet` mode the user should have an option to bail out if there is an error when resolving the diff.
> 
> Also: Probably, the resolving algorithm should be fixed so that it doesn't continue if there is an API error.

Indeed, perhaps it would be better to fail by default if api-erros
cannot be resolved. And rather continue with a --force flag if that
makes sense.

**Comment by canpolat** [comment]
This is related to #480 (probably a duplicate)

**Comment by canpolat** [comment]
When I read my initial comments, I see the ramblings of a confused man :)

I agree with @gauteh that it's better to have fail-by-default behavior.

Optionally, when failed, `drive` would ask the user if she wants to retry.

Optionally we could have a `retry-count` flag where the user can set the retry count (this can be in addition or instead of the internal retry mechanism) in such failures to ease automation (some people may want to retry indefinitely).

**Comment by odeke-em** [comment]
@gauteh and @canpolat, yap great plan but please don't forget a very large of users won't bother changing defaults as observed in interactions. I agree, adding a `--retry-count` flag is a great idea. If you wouldn't mind, could you please make a separate issue for this request?

**Comment by canpolat** [comment]
> I agree, adding a --retry-count flag is a great idea. If you wouldn't mind, could you please make a separate issue for this request?

#529 is created.

> (...) please don't forget a very large of users won't bother changing defaults as observed in interactions.

@odeke-em I'm not a big fan of changing defaults either, but continuing despite an API error doesn't sound correct to me (because we know that the diff will be wrong). My 2 cents :)

**Comment by odeke-em** [comment]
Cool, thanks for creating that issue.

Actually you might be suprised that a 5XX or even API error is the kind of error that is recommended for retry. I'll just give you a quick example, a server end point restart arriving at the same time as a request will error out give you a 5XX error. I've observed this quite a lot with Google Drive's backend. In fact on the Google Developers' page, they recommend implementing an exponential backoff(which is exactly what drive does) because 5XX errors are retryable. Please see https://developers.google.com/drive/v2/web/handle-errors

**Comment by gauteh** [comment]
Canpolat writes on December 24, 2015 15:53:
>> I agree, adding a --retry-count flag is a great idea. If you wouldn't mind, could you please make a separate issue for this request?
> 
> #529 is created.
> 
>> (...) please don't forget a very large of users won't bother changing defaults as observed in interactions.
> 
> @odeke-em I'm not a big fan of changing defaults either, but continuing despite an API error doesn't sound correct to me (because we know that the diff will be wrong). My 2 cents :)

As a new user I expected drive to quit on un-safe failures, I would
expect most users to want the same. 

If the error can be recovered from, by re-trying and it is possible to
continue and being certain that there will be no loss of data or
incomplete transfers then I don't consider that an un-safe failure.

**Comment by odeke-em** [comment]
"un-safe" failures. I don't know if you'd like your best-try connection to drop packets as soon as any problem that could be resolved was detected. If the max number of retries is reached, it also fails.

I have no opposition changing it back to just failing through but that's what drive was before and that wasn't reliable at all, in fact it was inconveniencing because if there were rate limits or temporary errors, users would then have to re-run drive again -- bad user experience. In the case of fails, as previously mentioned, this errors will be logged to the user and the files won't be uploaded but am sort of opposed to entirely failing on errors that are well documented as retryable. That's just my opinion, but it could be changed if you can convince me that the benefits outweigh the concerns of user experience and a frail like behavior.

**Comment by canpolat** [comment]
> Actually you might be surprised that a 5XX or even API error is the kind of error that is recommended for retry.

Just to clarify my point: I'm not against retries. Automatic retry upon API failure seems correct to me. But if all the retry attempts (exponential back-off) still fail, I think drive should stop with an error code instead of using the calculated diff and pushing/pulling stuff. Quoting from the original post:

    Resolving...
    googleapi: Error 503: Backend Error, backendError
    36163636 / 1284737695
    [===>-- ... --------------------------------------------------------] 2.81 % 37m11s

This should not happen (remember, here, in fact, local directory and cloud are in sync and `drive` thinks it is not probably due to the API error). This is my original point. So, in this regard, changing the default to me doesn't mean that drive should err out on the first failure, but that it should err out when retries fail. I don't think removing the retries is a good idea.

Then, on top of that, the user can add a `retry-count` that will retry to whole process (sorry for the confusion I caused). But this is just nice-to-have.

**Comment by CountryBumkin** [comment]
I suggest that 'fail safe' is important - especially for automated usage. My thoughts here... https://github.com/odeke-em/drive/issues/480#issuecomment-156981798

regards CB
